pulsar-client-kafka-compat/pulsar-client-kafka/src/main/java/org/apache/kafka/clients/producer/PulsarKafkaProducer.java

```diff
@@ -298,13 +303,17 @@ private int buildMessage(TypedMessageBuilder<byte[]> builder, ProducerRecord<K,
     }
 
     private String getKey(String topic, K key) {
-        // If key is a String, we can use it as it is, otherwise, serialize to byte[] and encode in base64
-        if (keySerializer instanceof StringSerializer) {
-            return (String) key;
-        } else {
-            byte[] keyBytes = keySerializer.serialize(topic, key);
-            return Base64.getEncoder().encodeToString(keyBytes);
+        if (keySerializer instanceof PulsarKafkaSchema) {
```

Comment:
the logic can be simplified with following logic. and also you can handle both StringSerializer and StringSchema.

```
if (key instanceof String) {
    return (String) key;
}

if (keySerializer instanceof PulsarKafkaSchema) {
    ((PulsarKafkaSchema) keySerializer).setTopic(topic);
}

byte[] keyBytes = keySerializer.encode(key);
return Base64.getEncoder().encodeToString(keyBytes);
```